payjoin/src/uri.rs

```diff
@@ -1,105 +1,190 @@
+        self.inner.extras.disable_output_substitution()
+    }
+    pub fn pj_is_supported(&self) -> bool { self.inner.extras.pj_is_supported() }
+    #[cfg(feature = "v2")]
+    pub fn ohttp_config(&self) -> Option<ohttp::KeyConfig> { self.inner.extras.ohttp_config() }
 }
 
-impl<'a> UriExt<'a> for Uri<'a, NetworkChecked> {
-    fn check_pj_supported(self) -> Result<PjUri<'a>, bip21::Uri<'a>> {
-        match self.extras {
-            Payjoin::Supported(payjoin) | Payjoin::V2Only(payjoin) => {
-                let mut uri = bip21::Uri::with_extras(self.address, payjoin);
-                uri.amount = self.amount;
-                uri.label = self.label;
-                uri.message = self.message;
+impl<'a> PayjoinUri<'a, NetworkChecked, Payjoin> {
+    pub fn with_extras(address: Address<NetworkChecked>, extras: Payjoin) -> Self {
+        Self { inner: bip21::Uri::with_extras(address, extras) }
+    }
+    #[cfg(feature = "v2")]
+    pub fn ohttp_config(&self) -> Option<ohttp::KeyConfig> { self.inner.extras.ohttp_config() }
+}
 
-                Ok(uri)
-            }
-            Payjoin::Unsupported => {
-                let mut uri = bip21::Uri::new(self.address);
-                uri.amount = self.amount;
-                uri.label = self.label;
-                uri.message = self.message;
+impl<'a> bip21::SerializeParams for &'a Payjoin {
+    type Key = &'static str;
+    type Value = String;
+    type Iterator = std::vec::IntoIter<(Self::Key, Self::Value)>;
 
-                Err(uri)
-            }
+    fn serialize_params(self) -> Self::Iterator {
+        match self {
+            Payjoin::Supported(params) => params.serialize_params(),
+            Payjoin::V2Only(params) => params.serialize_params(),
+            Payjoin::Unsupported => vec![].into_iter(),
         }
     }
 }
 
+impl<'a> bip21::SerializeParams for &'a PayjoinParams {
+    type Key = &'static str;
+    type Value = String;
+    type Iterator = std::vec::IntoIter<(Self::Key, Self::Value)>;
+
+    fn serialize_params(self) -> Self::Iterator {
+        vec![
+            ("pj", self.endpoint.as_str().to_string()),
+            ("pjos", if self.disable_output_substitution { "1" } else { "0" }.to_string()),
+            #[cfg(feature = "v2")]
+            // this unwrap is safe because we know that the config is required for v2
+            ("ohttp", encode_ohttp_config(self.ohttp_config.as_ref().unwrap()).unwrap_or_default()),
+        ]
+        .into_iter()
+    }
+}
+
 impl PayjoinParams {
     pub fn is_output_substitution_disabled(&self) -> bool { self.disable_output_substitution }
 }
 
+impl From<bip21::Uri<'_, NetworkChecked, PayjoinParams>>
+    for PayjoinUri<'_, NetworkChecked, Payjoin>
+{
+    fn from(value: bip21::Uri<'_, NetworkChecked, PayjoinParams>) -> Self {
+        let pj_uri = PjUriBuilder::new(
+            value.address,
+            value.extras.endpoint,
+            value.amount,
+            None,
```

Comment:
we should pass these params obviously, I just didnt wanna fight the conversion from Param to String yet